e2e/main_test.go

```diff
@@ -22,12 +21,11 @@ func TestMain(m *testing.M) {
 
 	flag.Parse()
 
-	logger, err := zap.NewDevelopment()
+	log, err := test.CreateLogger()
 	if err != nil {
 		panic(err)
 	}
-
-	log := logger.Sugar()
+	// TODO: Sync the log on exit
```

Comment:
It seems that the best way to handle the conflict between os.Exit() and defer is to call it only once in main() or in the case of tests, in TestMain(). This also means we cannot call any function that calls os.Exit() like log.Fatal().

We can do this like this:

```Go
func TestMain(m *testing.M) {
    os.Exit(run(m))
}

func run(m *testing.M) int {
    log := ...
    defer func() {
        _ = log.Sync()
    }()
    
    ...
    if err != nil {
        log.Errorf("error message... : %s", err)
        return 1
    }

    return m.Run()
}

With this we ensure that test failure or setup failure will return non-zero exit code, and defer function will run, and the code is very simple.

But lets no dalay this good change, we can improve this later.